network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/rekey/NetPRekeyer.kt

```diff
@@ -18,59 +18,118 @@ package com.duckduckgo.networkprotection.impl.rekey
-        if (currentTimeMillis - networkProtectionRepository.lastPrivateKeyUpdateTimeInMillis < TimeUnit.DAYS.toMillis(1)) {
+        val forceOrFalseInProductionBuilds = forceRekey.getAndResetValue()
+
+        val millisSinceLastKeyUpdate = System.currentTimeMillis() - networkProtectionRepository.lastPrivateKeyUpdateTimeInMillis
+        if (!forceOrFalseInProductionBuilds && millisSinceLastKeyUpdate < TimeUnit.DAYS.toMillis(1)) {
             logcat { "Less than 24h passed, skip re-keying" }
             return
         }
-        if (keyguardManager.deviceLocked()) {
-            networkProtectionRepository.privateKey = null
+
+        val newKey = runCatching {
+            val newKeys = KeyPair()
+            wgVpnControllerService.registerKey(RegisterKeyBody(newKeys.publicKey.toBase64(), "*"))
+
+            newKeys
+        }.onFailure {
+            logcat(LogPriority.ERROR) { "Failed registering the new key during re-keying: ${it.asLog()}" }
+        }.getOrNull()
+        if (newKey == null) {
+            return
+        } else {
+            logcat { "Re-keying with public key: ${newKey.publicKey.toBase64()}" }
+        }
+
+        if (deviceLockedChecker.invoke() || forceOrFalseInProductionBuilds) {
             if (vpnFeaturesRegistry.isFeatureRegistered(NetPVpnFeature.NETP_VPN)) {
                 logcat { "Restarting VPN after clearing client keys" }
+                networkProtectionPixels.reportRekeyCompleted()
+                networkProtectionRepository.privateKey = newKey.privateKey.toBase64()
                 vpnFeaturesRegistry.refreshFeature(NetPVpnFeature.NETP_VPN)
             } else {
-                logcat { "Cancelling scheduled rekey" }
-                workManager.cancelUniqueWork(DAILY_NETP_REKEY_TAG)
+                logcat(LogPriority.ERROR) { "Re-key work should not happen" }
             }
-            networkProtectionPixels.reportRekeyCompleted()
         } else {
             logcat { "Device not locked, skip re-keying" }
         }
     }
 
-    private fun KeyguardManager?.deviceLocked(): Boolean {
-        return this?.isKeyguardLocked == true
+    suspend fun forceRekey() {
+        if (appBuildConfig.isInternalBuild()) {
+            forceRekey.set(true)
+            doRekey()
+        } else {
+            logcat(LogPriority.ERROR) { "Force re-key not allowed in production builds" }
+        }
+    }
+}
+
+@Retention(AnnotationRetention.BINARY)
+@Qualifier
+private annotation class InternalApi
+
+// visible for testing
+internal typealias DeviceLockedChecker = () -> Boolean
+
+@Module
+@ContributesTo(VpnScope::class)
+object DeviceLockedCheckerModule {
+
+    @Provides
+    @InternalApi
+    fun provideDeviceLockChecker(context: Context): DeviceLockedChecker {
+        val keyguardManager = runCatching { context.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager }.getOrNull()
+
+        return {
+            keyguardManager?.isDeviceLocked ?: true
```

Comment:
“```
Returns whether the device is currently locked and requires a PIN, pattern or password to unlock.
Returns:
true if unlocking the device currently requires a PIN, pattern or password.
```”

Wouldn’t this return `false` if the screen is just closed and the user doesn’t have a lock screen?

Reply:
I will change the check to `(keyguardManager?.isDeviceLocked == true || powerManager?.isInteractive == false)`